docs/WASI-design-principles.md

```diff
@@ -0,0 +1,95 @@
+# WASI Design Principles
+
+## Capability-based security
+
+WASI is built using capability-based security principles. Access to external resources is always represented by *handles*, which are special values that are *unforgeable*, meaning there's no way to coerce an arbitrary integer or other type of value into a handle. WASI is also aiming to have no *ambient authorities*, meaning that there should be no way to request a handle purely by providing a string or other user-controlled identifier providing the name of a resource. With these two properties, the only ways to obtain access to resources are to be explicitly given handles, or to perform operations on handles which return new handles.
```

Comment:
`special values that are *unforgeable*`.. this isn't true today, which I think warrants mentioning here.

Reply:
It's sort of true, in that in C/etc. the handles are conceptually still there, they're just stored in tables for you. You can't forge new handles that aren't already in your table.

This is discussed more a few paragraphs later; does that clarify things?

Reply:
Yes, I agree that it does become more clear with the followup paragraphs.   I just found it slightly misleading to open with "special values that are *unforgeable*" since the current state is of WASI is that (as far as I understand) FDs are exactly as un-forgable is POSIX FDs.    You can conjure them out of thin air and use them (as long as your process you are in already has such as FD open).   For this reason I think the unforgability part might be better of explicitly in the a forward looking section of this document.

Reply:
We can still do shared-nothing linking though, even with libraries written in C, even when they use integer indices internally, because individual libraries can have their own tables. So it's true that C code could forge any index it wants, but it's still limited to accessing those things which have been placed in its table.

Reply:
I've now rewritten the paragraph that explains this, so that it now starts with "Integer indices are themselves forgeable...", which hopefully clarifies this.